language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/semantictokens/SemanticTokensVisitor.java

```diff
@@ -86,23 +87,32 @@
     // Sorted tree of semantic tokens
     private final Set<SemanticToken> semanticTokens;
     private final SemanticTokensContext semanticTokensContext;
+    private final String path;
+    private final SemanticModel semanticModel;
+
+    private Document document;
 
     public SemanticTokensVisitor(SemanticTokensContext semanticTokensContext) {
         this.semanticTokens = new TreeSet<>(SemanticToken.semanticTokenComparator);
         this.semanticTokensContext = semanticTokensContext;
+        Optional<String> optionalPath =
+                this.semanticTokensContext.workspace().relativePath(this.semanticTokensContext.filePath());
+        this.path = optionalPath.orElse(null);
+        Optional<SemanticModel> optionalSemanticModel = this.semanticTokensContext.currentSemanticModel();
+        this.semanticModel = optionalSemanticModel.orElse(null);
```

Comment:
> Also, you don't need to cache the semantic model for a given module since the operation context itself caches the semantic model. So we can avoid the caching here

We have been using `this.semanticTokensContext.currentSemanticModel()` in the visitor. But, according to the performance analysis, accessing the currentSemanticModel() takes a significant time when we access that again and again. Hence, moved to caching semantic model at the visitor level.

Reply:
Can you please re-check this? In the context we cache it. Please have a look at https://github.com/ballerina-platform/ballerina-lang/blob/master/language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/contexts/AbstractDocumentServiceContext.java#L220. 
So there will be no semantic model access per each call.

Reply:
Updated.